src/raw/mod.rs

```diff
@@ -0,0 +1,1634 @@
+
+                let id_size = 12;
+
+                if self.doc.data[valueoffset + string_size - 1] != 0 {
+                    return Some(Err(RawError::MalformedValue(
+                        "DBPointer string not null-terminated".into(),
+                    )));
+                }
+
+                string_size + id_size
+            }
+            ElementType::JavaScriptCode => {
+                let size =
+                    4 + i32_from_slice(&self.doc.data[valueoffset..valueoffset + 4]) as usize;
+
+                if self.doc.data[valueoffset + size - 1] != 0 {
+                    return Some(Err(RawError::MalformedValue(
+                        "javascript code not null-terminated".into(),
+                    )));
+                }
+
+                size
+            }
+            ElementType::Symbol => {
+                4 + i32_from_slice(&self.doc.data[valueoffset..valueoffset + 4]) as usize
+            }
+            ElementType::JavaScriptCodeWithScope => {
+                let size = i32_from_slice(&self.doc.data[valueoffset..valueoffset + 4]) as usize;
+
+                if self.doc.data[valueoffset + size - 1] != 0 {
+                    return Some(Err(RawError::MalformedValue(
+                        "javascript with scope not null-terminated".into(),
+                    )));
+                }
+
+                size
+            }
+            ElementType::Int32 => 4,
+            ElementType::Timestamp => 8,
+            ElementType::Int64 => 8,
+            ElementType::Decimal128 => 16,
+            ElementType::MaxKey => 0,
+            ElementType::MinKey => 0,
+        };
+
+        let nextoffset = valueoffset + element_size;
+        self.offset = nextoffset;
+
+        Some(Ok((
+            key,
+            elem::RawBson::new(element_type, &self.doc.data[valueoffset..nextoffset]),
+        )))
+    }
+}
+
+/// A BSON array referencing raw bytes stored elsewhere.
+pub struct RawArray {
+    doc: RawDocumentRef,
+}
+
+impl RawArray {
+    fn new(data: &[u8]) -> RawResult<&RawArray> {
+        Ok(RawArray::from_doc(RawDocumentRef::new(data)?))
+    }
+
+    fn from_doc(doc: &RawDocumentRef) -> &RawArray {
+        // SAFETY:
+        //
+        // Dereferencing a raw pointer requires unsafe due to the potential that the pointer is
+        // null, dangling, or misaligned. We know the pointer is not null or dangling due to the
+        // fact that it's created by a safe reference. Converting &RawDocumentRef to *const
+        // RawDocumentRef will be properly aligned due to them being references to the same type,
+        // and converting *const RawDocumentRef to *const RawArray is aligned due to the fact that
+        // the only field in a RawArray is a RawDocumentRef, meaning the structs are represented
+        // identically at the byte level.
+        unsafe { &*(doc as *const RawDocumentRef as *const RawArray) }
```

Comment:
i think this `unsafe` can similarly be avoided by making `RawDocumentRef` have a lifetime annotation.